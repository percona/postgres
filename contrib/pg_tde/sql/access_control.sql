CREATE EXTENSION IF NOT EXISTS pg_tde;

CREATE USER regress_pg_tde_access_control;

SET ROLE regress_pg_tde_access_control;

-- should throw access denied
SELECT pg_tde_add_key_provider_file('file-vault', '/tmp/pg_tde_test_keyring.per');
SELECT pg_tde_set_principal_key('test-db-principal-key', 'file-vault');

RESET ROLE;

SELECT pg_tde_grant_local_key_management_to_role('regress_pg_tde_access_control');
SELECT pg_tde_grant_key_viewer_to_role('regress_pg_tde_access_control');

SET ROLE regress_pg_tde_access_control;

-- should now be allowed
SELECT pg_tde_add_key_provider_file('file-vault', '/tmp/pg_tde_test_keyring.per');
SELECT pg_tde_add_key_provider_file('file-2', '/tmp/pg_tde_test_keyring_2.per');
SELECT pg_tde_set_principal_key('test-db-principal-key', 'file-vault');
SELECT * FROM pg_tde_list_all_key_providers();
SELECT principal_key_name, key_provider_name, key_provider_id FROM pg_tde_principal_key_info();

RESET ROLE;

SELECT pg_tde_revoke_key_viewer_from_role('regress_pg_tde_access_control');

SET ROLE regress_pg_tde_access_control;

-- verify the view access is revoked
SELECT * FROM pg_tde_list_all_key_providers();
SELECT principal_key_name, key_provider_name, key_provider_id FROM pg_tde_principal_key_info();

RESET ROLE;

DROP EXTENSION pg_tde CASCADE;
