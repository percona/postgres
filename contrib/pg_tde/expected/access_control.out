CREATE EXTENSION IF NOT EXISTS pg_tde;
CREATE USER regress_pg_tde_access_control;
SET ROLE regress_pg_tde_access_control;
-- should throw access denied
SELECT pg_tde_add_key_provider_file('file-vault', '/tmp/pg_tde_test_keyring.per');
ERROR:  permission denied for function pg_tde_add_key_provider_file
SELECT pg_tde_set_principal_key('test-db-principal-key', 'file-vault');
ERROR:  permission denied for function pg_tde_set_principal_key
RESET ROLE;
SELECT pg_tde_grant_local_key_management_to_role('regress_pg_tde_access_control');
 pg_tde_grant_local_key_management_to_role 
-------------------------------------------
 
(1 row)

SELECT pg_tde_grant_key_viewer_to_role('regress_pg_tde_access_control');
 pg_tde_grant_key_viewer_to_role 
---------------------------------
 
(1 row)

SET ROLE regress_pg_tde_access_control;
-- should now be allowed
SELECT pg_tde_add_key_provider_file('file-vault', '/tmp/pg_tde_test_keyring.per');
 pg_tde_add_key_provider_file 
------------------------------
                            1
(1 row)

SELECT pg_tde_add_key_provider_file('file-2', '/tmp/pg_tde_test_keyring_2.per');
 pg_tde_add_key_provider_file 
------------------------------
                            2
(1 row)

SELECT pg_tde_set_principal_key('test-db-principal-key', 'file-vault');
 pg_tde_set_principal_key 
--------------------------
 t
(1 row)

SELECT * FROM pg_tde_list_all_key_providers();
 id | provider_name | provider_type |                           options                            
----+---------------+---------------+--------------------------------------------------------------
  1 | file-vault    | file          | {"type" : "file", "path" : "/tmp/pg_tde_test_keyring.per"}
  2 | file-2        | file          | {"type" : "file", "path" : "/tmp/pg_tde_test_keyring_2.per"}
(2 rows)

SELECT principal_key_name, key_provider_name, key_provider_id FROM pg_tde_principal_key_info();
  principal_key_name   | key_provider_name | key_provider_id 
-----------------------+-------------------+-----------------
 test-db-principal-key | file-vault        |               1
(1 row)

RESET ROLE;
SELECT pg_tde_revoke_key_viewer_from_role('regress_pg_tde_access_control');
 pg_tde_revoke_key_viewer_from_role 
------------------------------------
 
(1 row)

SET ROLE regress_pg_tde_access_control;
-- verify the view access is revoked
SELECT * FROM pg_tde_list_all_key_providers();
ERROR:  permission denied for function pg_tde_list_all_key_providers
SELECT principal_key_name, key_provider_name, key_provider_id FROM pg_tde_principal_key_info();
ERROR:  permission denied for function pg_tde_principal_key_info
RESET ROLE;
DROP EXTENSION pg_tde CASCADE;
